File: packages/ui/src/standards/symbol-generators/cabinet.ts
import type { CabinetElement } from '../../services/kernel-bridge'
import type { SymbolGenerationOptions, SymbolPrimitiveBundle } from '../symbol-primitives'
import {
  addRect, addLine, addText, addEllipse, clamp,
  type SymbolPrimitive,
} from './_helpers'

export function generateCabinetPrimitives(
  element: CabinetElement,
  options: SymbolGenerationOptions,
): SymbolPrimitiveBundle {
  const [x, y] = element.position
  const rotation = element.rotation ?? 0
  const width = Math.max(0.25, Math.abs(element.width))
  const depth = Math.max(0.25, Math.abs(element.depth))
  const hw = width / 2
  const hd = depth / 2
  const minDim = Math.min(width, depth)

  const primitives: SymbolPrimitive[] = []

  const isAboveCut =
    element.cabinet_type === 'upper' ||
    element.cabinet_type === 'corner_upper' ||
    element.cabinet_type === 'appliance_garage'

  switch (element.cabinet_type) {
    case 'base':
    case 'drawer_base': {
      // Solid rectangle + countertop overhang line at front + door partitions
      addRect(primitives, [x, y], rotation, width, depth, 'primary')
      // Countertop overhang at front
      addLine(primitives, [x, y], rotation, [-hw, hd + 0.025], [hw, hd + 0.025], 'secondary')
      // Door partition for 2-door cabinets
      if (element.door_count >= 2) {
        addLine(primitives, [x, y], rotation, [0, -hd], [0, hd], 'secondary')
      }
      // Horizontal drawer lines for drawer_base
      if (element.cabinet_type === 'drawer_base') {
        for (let i = 1; i <= 3; i++) {
          const ly = -hd + (depth * i / 4)
          addLine(primitives, [x, y], rotation, [-hw, ly], [hw, ly], 'secondary')
        }
      }
      break
    }

    case 'upper': {
      // Dashed rectangle (above cut plane) + dashed shelf line
      addRect(primitives, [x, y], rotation, width, depth, 'primary', [0, 0], true)
      addLine(primitives, [x, y], rotation, [-hw, 0], [hw, 0], 'secondary', true)
      break
    }

    case 'tall':
    case 'pantry': {
      // Solid rectangle + horizontal line separating upper/lower
      addRect(primitives, [x, y], rotation, width, depth, 'primary')
      addLine(primitives, [x, y], rotation, [-hw, 0], [hw, 0], 'secondary')
      break
    }

    case 'corner_base': {
      // Rectangle with diagonal line
      addRect(primitives, [x, y], rotation, width, depth, 'primary')
      addLine(primitives, [x, y], rotation, [-hw, hd], [hw, -hd], 'secondary')
      break
    }

    case 'corner_upper': {
      // Dashed rectangle with dashed diagonal
      addRect(primitives, [x, y], rotation, width, depth, 'primary', [0, 0], true)
      addLine(primitives, [x, y], rotation, [-hw, hd], [hw, -hd], 'secondary', true)
      break
    }

    case 'corner_tall': {
      // Solid outline with diagonal + horizontal divider
      addRect(primitives, [x, y], rotation, width, depth, 'primary')
      addLine(primitives, [x, y], rotation, [-hw, hd], [hw, -hd], 'secondary')
      addLine(primitives, [x, y], rotation, [-hw, 0], [hw, 0], 'secondary')
      break
    }

    case 'sink_base': {
      // Rectangle + countertop overhang + sink basin oval inside
      addRect(primitives, [x, y], rotation, width, depth, 'primary')
      addLine(primitives, [x, y], rotation, [-hw, hd + 0.025], [hw, hd + 0.025], 'secondary')
      addEllipse(primitives, [x, y], rotation, [0, 0.05], width * 0.3, depth * 0.25, 'secondary')
      break
    }

    case 'lazy_susan': {
      // Square with X diagonals to indicate rotation
      addRect(primitives, [x, y], rotation, width, depth, 'primary')
      addLine(primitives, [x, y], rotation, [-hw, -hd], [hw, hd], 'secondary')
      addLine(primitives, [x, y], rotation, [hw, -hd], [-hw, hd], 'secondary')
      break
    }

    case 'blind_corner': {
      // Rectangle with vertical partition
      addRect(primitives, [x, y], rotation, width, depth, 'primary')
      addLine(primitives, [x, y], rotation, [0, -hd], [0, hd], 'secondary')
      break
    }

    case 'appliance_garage': {
      // Dashed rectangle (mounted above counter)
      addRect(primitives, [x, y], rotation, width, depth, 'primary', [0, 0], true)
      break
    }

    default:
      addRect(primitives, [x, y], rotation, width, depth, 'primary', [0, 0], isAboveCut)
      break
  }

  // Label
  if (options.showLabels) {
    const label = element.cabinet_type.replace(/_/g, ' ').toUpperCase()
    addText(primitives, [x, y], rotation, label, clamp(minDim * 0.22, 0.06, 0.14))
  }

  return {
    domain: 'cabinet',
    primitives,
  }
}

File: crates/bcad-2d/src/symbols/cabinet.rs
//! Cabinet symbol generators.
//!
//! Geometry closely mirrors the TypeScript generators in
//! `packages/ui/src/standards/symbol-generators/cabinet.ts`.

use crate::style::{DomainColor, StyledPrimitive, SymbolLineClass};

use super::{add_ellipse, add_line, add_rect};

const DOMAIN: DomainColor = DomainColor::Cabinet;

/// Base cabinet: solid rect + countertop overhang + optional door partition.
pub fn base_cabinet(
    center: [f64; 2],
    rotation: f64,
    width: f64,
    depth: f64,
    door_count: u32,
) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    let hw = width / 2.0;
    let hd = depth / 2.0;
    add_rect(
        &mut out,
        DOMAIN,
        center,
        rotation,
        width,
        depth,
        SymbolLineClass::Primary,
        [0.0, 0.0],
        false,
    );
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [-hw, hd + 0.025],
        [hw, hd + 0.025],
        SymbolLineClass::Secondary,
        false,
    );
    if door_count >= 2 {
        add_line(
            &mut out,
            DOMAIN,
            center,
            rotation,
            [0.0, -hd],
            [0.0, hd],
            SymbolLineClass::Secondary,
            false,
        );
    }
    out
}

/// Drawer base: base cabinet + horizontal drawer lines.
pub fn drawer_base(
    center: [f64; 2],
    rotation: f64,
    width: f64,
    depth: f64,
) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    let hw = width / 2.0;
    let hd = depth / 2.0;
    add_rect(
        &mut out,
        DOMAIN,
        center,
        rotation,
        width,
        depth,
        SymbolLineClass::Primary,
        [0.0, 0.0],
        false,
    );
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [-hw, hd + 0.025],
        [hw, hd + 0.025],
        SymbolLineClass::Secondary,
        false,
    );
    for i in 1..=3 {
        let ly = -hd + (depth * i as f64 / 4.0);
        add_line(
            &mut out,
            DOMAIN,
            center,
            rotation,
            [-hw, ly],
            [hw, ly],
            SymbolLineClass::Secondary,
            false,
        );
    }
    out
}

/// Upper (wall) cabinet: dashed rectangle + dashed shelf line.
pub fn upper_cabinet(
    center: [f64; 2],
    rotation: f64,
    width: f64,
    depth: f64,
) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    let hw = width / 2.0;
    add_rect(
        &mut out,
        DOMAIN,
        center,
        rotation,
        width,
        depth,
        SymbolLineClass::Primary,
        [0.0, 0.0],
        true,
    );
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [-hw, 0.0],
        [hw, 0.0],
        SymbolLineClass::Secondary,
        true,
    );
    out
}

/// Tall / pantry cabinet: solid rect + horizontal divider.
pub fn tall_cabinet(
    center: [f64; 2],
    rotation: f64,
    width: f64,
    depth: f64,
) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    let hw = width / 2.0;
    add_rect(
        &mut out,
        DOMAIN,
        center,
        rotation,
        width,
        depth,
        SymbolLineClass::Primary,
        [0.0, 0.0],
        false,
    );
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [-hw, 0.0],
        [hw, 0.0],
        SymbolLineClass::Secondary,
        false,
    );
    out
}

/// Corner base cabinet: rect + diagonal line.
pub fn corner_base(
    center: [f64; 2],
    rotation: f64,
    width: f64,
    depth: f64,
) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    let hw = width / 2.0;
    let hd = depth / 2.0;
    add_rect(
        &mut out,
        DOMAIN,
        center,
        rotation,
        width,
        depth,
        SymbolLineClass::Primary,
        [0.0, 0.0],
        false,
    );
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [-hw, hd],
        [hw, -hd],
        SymbolLineClass::Secondary,
        false,
    );
    out
}

/// Corner upper cabinet: dashed rect + dashed diagonal.
pub fn corner_upper(
    center: [f64; 2],
    rotation: f64,
    width: f64,
    depth: f64,
) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    let hw = width / 2.0;
    let hd = depth / 2.0;
    add_rect(
        &mut out,
        DOMAIN,
        center,
        rotation,
        width,
        depth,
        SymbolLineClass::Primary,
        [0.0, 0.0],
        true,
    );
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [-hw, hd],
        [hw, -hd],
        SymbolLineClass::Secondary,
        true,
    );
    out
}

/// Corner tall cabinet: solid rect + diagonal + horizontal divider.
pub fn corner_tall(
    center: [f64; 2],
    rotation: f64,
    width: f64,
    depth: f64,
) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    let hw = width / 2.0;
    let hd = depth / 2.0;
    add_rect(
        &mut out,
        DOMAIN,
        center,
        rotation,
        width,
        depth,
        SymbolLineClass::Primary,
        [0.0, 0.0],
        false,
    );
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [-hw, hd],
        [hw, -hd],
        SymbolLineClass::Secondary,
        false,
    );
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [-hw, 0.0],
        [hw, 0.0],
        SymbolLineClass::Secondary,
        false,
    );
    out
}

/// Sink base cabinet: rect + countertop overhang + basin ellipse.
pub fn sink_base(center: [f64; 2], rotation: f64, width: f64, depth: f64) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    let hw = width / 2.0;
    let hd = depth / 2.0;
    add_rect(
        &mut out,
        DOMAIN,
        center,
        rotation,
        width,
        depth,
        SymbolLineClass::Primary,
        [0.0, 0.0],
        false,
    );
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [-hw, hd + 0.025],
        [hw, hd + 0.025],
        SymbolLineClass::Secondary,
        false,
    );
    add_ellipse(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [0.0, 0.05],
        width * 0.3,
        depth * 0.25,
        SymbolLineClass::Secondary,
        24,
    );
    out
}

/// Lazy susan: square + X diagonals.
pub fn lazy_susan(center: [f64; 2], rotation: f64, width: f64, depth: f64) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    let hw = width / 2.0;
    let hd = depth / 2.0;
    add_rect(
        &mut out,
        DOMAIN,
        center,
        rotation,
        width,
        depth,
        SymbolLineClass::Primary,
        [0.0, 0.0],
        false,
    );
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [-hw, -hd],
        [hw, hd],
        SymbolLineClass::Secondary,
        false,
    );
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [hw, -hd],
        [-hw, hd],
        SymbolLineClass::Secondary,
        false,
    );
    out
}

/// Blind corner: rect + vertical partition.
pub fn blind_corner(
    center: [f64; 2],
    rotation: f64,
    width: f64,
    depth: f64,
) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    let hd = depth / 2.0;
    add_rect(
        &mut out,
        DOMAIN,
        center,
        rotation,
        width,
        depth,
        SymbolLineClass::Primary,
        [0.0, 0.0],
        false,
    );
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [0.0, -hd],
        [0.0, hd],
        SymbolLineClass::Secondary,
        false,
    );
    out
}

/// Appliance garage: dashed rectangle (above counter).
pub fn appliance_garage(
    center: [f64; 2],
    rotation: f64,
    width: f64,
    depth: f64,
) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    add_rect(
        &mut out,
        DOMAIN,
        center,
        rotation,
        width,
        depth,
        SymbolLineClass::Primary,
        [0.0, 0.0],
        true,
    );
    out
}

/// Pantry cabinet (same as tall_cabinet).
pub fn pantry(center: [f64; 2], rotation: f64, width: f64, depth: f64) -> Vec<StyledPrimitive> {
    tall_cabinet(center, rotation, width, depth)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn base_cabinet_produces_primitives() {
        let prims = base_cabinet([0.0, 0.0], 0.0, 0.6, 0.6, 2);
        // rect + overhang + partition = 3
        assert!(prims.len() >= 3);
    }

    #[test]
    fn upper_cabinet_is_dashed() {
        let prims = upper_cabinet([0.0, 0.0], 0.0, 0.6, 0.3);
        assert!(prims.len() >= 2);
        assert!(prims[0].dashed, "upper cabinet rect should be dashed");
    }

    #[test]
    fn lazy_susan_has_x_diagonals() {
        let prims = lazy_susan([0.0, 0.0], 0.0, 0.6, 0.6);
        // rect + 2 diags = 3
        assert!(prims.len() >= 3);
    }
}
